2630 some support for data statement

hiker opened the pull request:
Adds some support for Data statements - they are turned into `UnsupportedFortranType`s, but this means that the rest of files containing data statements are properly handled now.

hiker commented:
I've also updated the dev guide (thanks for pointing this out in #2658)

arporter commented on doc/developer_guide/psyir_symbols.rst:
```diff
@@ -105,6 +105,9 @@ explicitly listed may be assumed to be unsupported):
 +----------------------+--------------------+--------------------+
 |                      |PUBLIC, PRIVATE     |                    |
 +----------------------+--------------------+--------------------+
+|                      |Data statements     |                    |
```
Thanks for updating this. Since I couldn't remember what data statements were and had to google it, perhaps they should be included in the "initialisation expressions" cell above, e.g. "Initialisation expressions (incl. limited support for `data` statements)"

hiker replied:
OK, I thought I had added it to that section (though I now realised that below are other non-init statements). TBH, I found that table a bit confusion. For now I have tried to move the 'Initialisation Expressions' to the left column, and then data statements to middle column next to it. Is that how you intended this to be?

arporter replied:
Sort of. It was supposed to cover initialisation in variable declarations too although I agree it wasn't very clear. Perhaps we should add "explicit initialisation" (as that's what the standard calls it) to the "supported" cell too?

arporter commented on src/psyclone/psyir/frontend/fparser2.py:
```diff
@@ -2626,6 +2626,12 @@ def process_declarations(self, parent, nodes, arg_list,
                     datatype=UnsupportedFortranType(str(node)))
+
+            elif isinstance(node, Fortran2003.Data_Stmt):
+                parent.symbol_table.new_symbol(
```
Comment please :-) c.f. Namelist_Stmt above.

arporter replied:
If you could include an example with an implicit loop here as justification/explanation of this approach then that would be great.

hiker replied:
Done (well, in the description of the new function I have added)

arporter commented on src/psyclone/tests/psyir/frontend/fparser2_data_statement_test.py:
```diff
@@ -0,0 +1,128 @@
+
+@pytest.mark.usefixtures("f2008_parser")
+def test_multiple_data_statement():
+    ''' Test that data statements are correctly captured. '''
```
"...that multiple data..."

hiker replied:
Done.

arporter commented on src/psyclone/tests/psyir/frontend/fparser2_data_statement_test.py:
```diff
@@ -0,0 +1,128 @@
+    symtab = routine.symbol_table
+    processor = Fparser2Reader()
+
+    # Test with a single namelist
```
s/namelist/data statement/ :-)

arporter replied:
...with multiple initialisations.

hiker replied:
Done and done ... this gives the totally wrong impression I copied that test from somewhere :)

arporter commented on src/psyclone/tests/psyir/frontend/fparser2_data_statement_test.py:
```diff
@@ -0,0 +1,128 @@
+
+@pytest.mark.usefixtures("f2008_parser")
+def test_data_statement_implicit_loop():
+    ''' Test that z data statement with an implicit loops is correctly
```
"that a data.." and "loop is correctly..." ?

hiker replied:
Done.

arporter commented on src/psyclone/tests/psyir/frontend/fparser2_data_statement_test.py:
```diff
@@ -0,0 +1,128 @@
+
+    # Test with a single namelist
+    reader = FortranStringReader(
+        "DATA (es(ies),ies=    0, 5) / 0.966483e-02,0.966483e-02,"
```
Oh yes. This makes it very clear why we only have partial support :-) (I was starting to think that it would be easy to convert into a canonical form. Although, having said that, it shouldn't be *too* difficult to add suitable statements. Not something for this PR though.)

hiker replied:
You are aware implied do-loops can be nested?  :)

arporter replied:
"walk in the park" :-)

arporter reviewed (changes requested):
Thanks for this Joerg, I think it's a good first step. I am wondering whether we should take note of:
![image](https://github.com/user-attachments/assets/354e1805-6e17-4d56-a500-5943df881e0e)
i.e. that any symbols initialised in a DATA statement effectively have the SAVE attribute. This is true for both Fortran2003 and 2008. I'm not sure about earlier standards.
I'll check integration tests next time. Appropriate doc has been updated.

hiker commented:
> Thanks for this Joerg, I think it's a good first step. I am wondering whether we should take note of: 
...
> i.e. that any symbols initialised in a DATA statement effectively have the SAVE attribute. This is true for both Fortran2003 and 2008. I'm not sure about earlier standards. I'll check integration tests next time. Appropriate doc has been updated.

OK, I have added this, including support for nested do loops, and derived statements, and added tests based on the F90 standard. You know fparser better than me, I hope I got the possible node type for variables correct (there seems to be a bit of a difference between fparser and the standard: 
```
R526:  data-stmt-object is variable
                        or data-implied-do
```
I assume the `variable` is skipped? So I hope I capture all possible ways a variable might be specified. Similar for:
```
R528 data-i-do-object is array-element
                      or scalar-structure-component
                      or data-implied-do
```
I don't see array-element or scalar-structure component, so I search for PartRef.

I've triggered CI.

arporter commented on src/psyclone/tests/psyir/frontend/fparser2_data_statement_test.py:
```diff
@@ -35,94 +35,188 @@
+               integer :: a, b
+            end type mytype
+            type(mytype) :: t
+            data t%a, t%b / 1, 2 /
```
Blimey, more Fortran I've never seen before.

arporter reviewed (changes requested):
Thanks Joerg, almost there now.
I've triggered the integration tests.

hiker commented:
All done, CI triggered and it finished successfully.

arporter reviewed (approved):
All requested changes have been made, thanks.
Integration tests are all fine.
Ref. guide builds fine.
NEMO performance is unaffected.
